api/src/main/java/org/eclipse/microprofile/config/Config.java

```diff
@@ -132,6 +132,17 @@
     /**
      * Return all property names returned by any of the underlying {@linkplain ConfigSource configuration sources}.
      * The order of the returned property names is unspecified.
+     * <p>
+     * The returned property names are unique; that is, if a name is returned once, it will not be returned again.
+     * The names returned may reflect a subset of the total set of available names.
```

Comment:
"names returned may reflect"?  Or "set of names must be"?

Reply:
The returned value is not a set, so we can't require it to be one.

Reply:
(But the notional set of "all property names" is a `Set`, yes?  So mustn't the `Iterable` here have `Set` semantics?)

My larger point was the use of "may": by this specification, I could return an `Iterable` with `nonexisting.property.name1, nonexisting.property.name2` in it when none of my `ConfigSource`s have such property names and I'd be compliant.

I also just noticed that I don't understand what "all property names" at the top means, given line 137.  Can I return a subset of the `ConfigSource` property names, and only a subset of them?  Or must I return all of them?  May I return values in the `Iterable` that do not have corresponding values in the property names originating from the `ConfigSource` instances?  Must I not?  You see what I'm getting at.

Reply:
"Return all property names returned by any of the underlying..." is unambiguous: if the underlying config source returns the property name, it must be reported by the iterable.  That said, it might be a good idea to loosen this a bit.

The word "may" means that the implementation might show a reduced set of names compared to what might naively be considered available.  This follows on to the normative language in `ConfigSource` which says "The returned property names may be a subset of the names of the total set of retrievable properties in this config source".  But I can tighten up that language.

> My larger point was the use of "may": by this specification, I could return an `Iterable` with `nonexisting.property.name1, nonexisting.property.name2` in it when none of my `ConfigSource`s have such property names and I'd be compliant.

That is not true; the returned names would not be a subset of the total available names in this case.  That said though, there's probably no reason to make this quite so strict: really the config object could be allowed to return any old set of names.

Maybe this would be better language:

```java
/**
 * Returns a sequence of configuration property names.
 * The order of the returned property names is unspecified.
 * <p>
 * The returned property names are unique; that is, if a name is returned once by a given iteration, it will not be returned again during that same iteration.
 * <p>
 * There is no guarantee about the completeness or currency of the names returned, nor is there any guarantee that a name that is returned by the iterator will resolve to a non-empty value or be found in any configuration source associated with the configuration.
* <p>
 * It is implementation-defined whether the returned names reflect a point-in-time "snapshot" of names, or
 * an aggregation of multiple point-in-time "snapshots", or a more dynamic view of the available property names.
 * Implementations are not required to return the same sequence of names on each iteration; however, the produced
 * {@link java.util.Iterator Iterator} must adhere to the contract of that class, and must not return any more
 * elements once its {@link java.util.Iterator#hasNext() hasNext()} method returns {@code false}.
 * <p>
 * The returned instance is thread safe and may be iterated concurrently.  The individual iterators are not thread-safe.
 */
```

Reply:
Sources `A`, `B` and `C` report `a`, `b`, and `c`.
The spec says: `[The returned Iterable must] return all property names returned by any of the underlying {@linkplain ConfigSource configuration sources}.`
That says I MUST return `a`, `b` and `c` if they are reported at the moment of this method's invocation.  OK.  They were (in this example).  Cool.  I'll (necessarily) add `a`, `b` and `c` to my `Iterable` as I'm building it.
Then the spec says: `The names returned may [or may not, my emphasis] reflect a subset of the total set of available names.`  Cool.  It doesn't say MUST.  I'll add `d` to my iterable for arcane implementation reasons.
With the new language (`There is no guarantee about the completeness or currency of the names returned, nor is there any guarantee that a name that is returned by the iterator will resolve to a non-empty value or be found in any configuration source associated with the configuration.`), `{ a, b, c, d }` is fine, as is `{ q, x, r }`.  At any point, it might also change during iteration to, say, `{ j, k, q }`.
Do I have this right so far?  Given how loose the method now can be, is it still worthwhile?

Reply:
> Sources `A`, `B` and `C` report `a`, `b`, and `c`.
> The spec says: `[The returned Iterable must] return all property names returned by any of the underlying {@linkplain ConfigSource configuration sources}.`
> That says I MUST return `a`, `b` and `c` if they are reported at the moment of this method's invocation. OK. They were (in this example). Cool. I'll (necessarily) add `a`, `b` and `c` to my `Iterable` as I'm building it.

I removed this language in the above proposal.

> Then the spec says: `The names returned may [or may not, my emphasis] reflect a subset of the total set of available names.` Cool. It doesn't say MUST. I'll add `d` to my iterable for arcane implementation reasons.

I removed this language as well.

> With the new language (`There is no guarantee about the completeness or currency of the names returned, nor is there any guarantee that a name that is returned by the iterator will resolve to a non-empty value or be found in any configuration source associated with the configuration.`), `{ a, b, c, d }` is fine, as is `{ q, x, r }`. At any point, it might also change during iteration to, say, `{ j, k, q }`.

Correct.  In Quarkus we inject names in the configuration so that configuration groups are read.  We remove names from the configuration if they correspond to sensitive configuration properties.  The proposed configuration interceptors (forthcoming) certainly allows all this and more.

> Do I have this right so far? Given how loose the method now can be, is it still worthwhile?

I think it is.  If you want, I could add a recommendation like "The implementation SHOULD return a set of names that is useful to a user that wishes to browse the configuration."  It would give a general intent without restricting application-specific requirements.

Sometimes it's OK for a specification to give latitude to the implementation.

Reply:
Goodness, yes, it is _frequently_ OK for a specification to give latitude to the implementation.  The delicate dance is: what's specified?  What's not?  In this case, the spec is now becoming "return whatever you want, so long as someone can iterate over it concurrently".  I just want to make sure that's the intent of this method.  If it _is_ the intent of this method, then yes, the "should" statement—giving a sense of the use cases involved—would be a good thing IMHO, since otherwise a method that lets me just return `frob` and `quux` for a hard-to-discern purpose is probably surface area we don't want to support. 😄

Reply:
Pushed.